Added pyserial 3.x support on Windows

altendky opened the pull request:
https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Added pyserial 3.x support on Windows

https://twistedmatrix.com/trac/ticket/8159

altendky commented:
~~Unintended push to wrong branch.~~

glyph committed:
Merge branch 'trunk' into 8159-altendky-support_pyserial_3

glyph commented:
(Code review is over on Trac - https://twistedmatrix.com/trac/ticket/8159#comment:14)

altendky committed:
Add handle access coverage tests for _win32serialport

altendky committed:
Also override Serial._reconfigure_port()

altendky committed:
Merge branch 'trunk' into 8159-altendky-support_pyserial_3

altendky committed:
Improve coverage for internet._win32serialport

altendky committed:
Merge branch 'trunk' into 8159-altendky-support_pyserial_3

altendky committed:
Test both pySerial 2.7 and latest on AppVeyor

https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Remove accidentally committed comment

https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Improve internet._win32serialport coverage (slightly)

https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Remove unneeded serial read if

From what I can tell, in any case that `win32file.GetOverlappedResult()`
would otherwise return zero it instead converts it into an exception.
As such, no need for the `if n:` check.  It will always be true.

https://github.com/pywin32/pypiwin32/issues/64

https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Merge branch 'trunk' into 8159-altendky-support_pyserial_3

altendky committed:
Merge branch 'trunk' into 8159-altendky-support_pyserial_3

altendky committed:
Merge branch 'trunk' into 8159-altendky-support_pyserial_3

altendky committed:
Merge branch 'trunk' into 8159-altendky-support_pyserial_3

exarkun commented on src/twisted/internet/_win32serialport.py:
```diff
@@ -56,40 +59,54 @@ def __init__(self, protocol, deviceNameOrPortNumber, reactor,
+    @property
+    def _serialHandle(self):
+        # Try to detect pyserial version to support internal API changes
+        if hasattr(self._serial, '_port_handle'):
```
`hasattr` can mask unexpected exceptions.  I'd suggest something like the following here:

```
serial = self._serial
try:
    # >= 3.0
    return serial._port_handle 
except AttributeError:
    # <= 2.7
    return serial.hComPort
```

exarkun commented on src/twisted/internet/test/test_win32serialport.py:
```diff
@@ -0,0 +1,137 @@
+        self.assertEqual(0,                   kwargs["xonxoff"])
+        self.assertEqual(0,                   kwargs["rtscts"])
+        self.assertEqual(None,                kwargs["timeout"])
+        port.connectionLost(reason='Cleanup')
```
`reason` should be a `Failure`.  The cheapest technically-correct way to do this is `connectionLost(Failure(Exception("Cleanup")))`.

altendky committed:
Merge remote-tracking branch 'upstream/trunk' into 8159-altendky-support_pyserial_3

altendky committed:
Try/except instead of hasattr() for pySerial version identification

https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Pass an actual Failure to connectionLost in serial tests

https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Merge branch 'trunk' into 8159-altendky-support_pyserial_3

altendky committed:
Merge branch 'trunk' into 8159-altendky-support_pyserial_3

altendky committed:
Avoid frequent (caught) exceptions detecting pySerial version

Only detect the pySerial version once per object in __init__() and then use the detected value in the _serialHandle property.

https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Fix missing self._ in self._serial references

https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Drop pySerial 2.x support

https://twistedmatrix.com/trac/ticket/8159

altendky committed:
Correct setup tests for 'pyserial >= 3.0'

https://twistedmatrix.com/trac/ticket/8159

altendky commented:
@hawkowl, I believe I made the changes you wanted.  Now requires pySerial >= 3.0 in _setup.py and on Windows only supports 3.x.  That is a point to be made, I am only aware of this incompatibility between 2.x and 3.x affecting Windows so the >=3.0 may be more strict than it needs to be on other platforms.

glyph commented:
Thank you @altendky for keeping up with all these review changes!
